addons/hr_expense/hr_expense.py

```diff
@@ -65,257 +32,219 @@ def _get_currency(self, cr, uid, context=None):
+    company_id = fields.Many2one('res.company', string='Company', required=True, readonly=True, states={'draft': [('readonly', False)], 'confirm':[('readonly', False)]}, default=lambda self: self.env.user.company_id)
+    state = fields.Selection([
+        ('draft', 'New'),
+        ('cancelled', 'Refused'),
+        ('confirm', 'Waiting Approval'),
+        ('accepted', 'Approved'),
+        ('done', 'Waiting Payment'),
+        ('paid', 'Paid'),
+        ],
+        string='Status', readonly=True, track_visibility='onchange', copy=False, default='draft', required=True,
+        help='When the expense request is created the status is \'Draft\'.\n It is confirmed by the user and request is sent to admin, the status is \'Waiting Confirmation\'.\
+        \nIf the admin accepts it, the status is \'Accepted\'.\n If the accounting entries are made for the expense request, the status is \'Waiting Payment\'.')
+
+    @api.multi
+    def unlink(self):
+        for sheet in self:
+            if sheet.state != 'draft':
+                raise Warning(_('You can only delete draft expenses!'))
+        return super(hr_expense_sheet, self).unlink()
+
+    @api.onchange('currency_id')
+    def onchange_currency_id(self):
+        account_journal = self.env['account.journal'].search([('type', '=', 'purchase'), ('currency', '=', self.currency_id.id), ('company_id', '=', self.company_id.id)], limit=1)
+        self.journal_id = account_journal.id
+
+    @api.multi
+    def sheet_draft(self):
+        for sheet in self:
+            sheet.line_ids.expense_line_draft()
+        return self.write({'state': 'draft'})
+
+    @api.multi
+    def sheet_confirm(self):
+        for sheet in self:
+            if not sheet.line_ids:
+                raise Warning(_('You cannot submit expense which has no expense line.'))
+            if sheet.employee_id and sheet.employee_id.parent_id.user_id:
+                self.message_subscribe_users(user_ids = sheet.employee_id.parent_id.user_id.ids)
+            if False not in [True if expense.state in ['draft'] else False for expense in sheet.line_ids]:
+                sheet.line_ids.write({'state': 'confirm'})
+        return self.write({'date_confirm': fields.Date.context_today(self)})
+
+    @api.multi
+    def sheet_accept(self):
+        for expense in self.line_ids:
+            if expense.state != 'cancel':
+                expense.expense_approved()
+        return self.write({'state': 'accepted', 'date_valid': fields.Date.context_today(self), 'user_valid': self._uid})
+
+    @api.multi
+    def sheet_approved(self):
+        for sheet in self:
+            if sheet.state in ['accepted', 'done', 'paid']:
+                raise Warning(_('You have already approved sheet!'))
+            if sheet.state in ['cancelled']:
+                sheet.signal_workflow('draft')
+            sheet.signal_workflow('confirm')
+            sheet.signal_workflow('validate')
```

Comment:
Don't know what are you trying to do here ???
It seems that re-factoring of workflow method is necessary.